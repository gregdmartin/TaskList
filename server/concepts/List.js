class List {
    constructor() {
        this._title = 'this is a list';
        this._items = [];
    }
    set title(newTitle) {
        this._title = newTitle;
    }
    get title() {
        return this._title;
    }
    get items() {
        return this._items;
    }
}

module.exports = List;
