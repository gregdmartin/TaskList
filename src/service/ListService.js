import Ajax from '../common/Ajax';

export const getLists = async () => {
  return Ajax.fetchAjax('api/listing/list?asXhr=true');
};
