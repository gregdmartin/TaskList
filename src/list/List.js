import React from 'react';
import Item from './Item';
import { createItem } from '../service/ListService';

/**
 * This is a UI list, which is mostly to use a collection of Items.
 */
class List extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      items: this.props.items || [],
    };
    this.createInput = React.createRef();
    this.newItemSubmit = this.newItemSubmit.bind(this);
    this.onEnter = this.onEnter.bind(this);
  }

  async onEnter(e) {
    if(e.key === 'Enter') {
      return this.newItemSubmit();
    }
  }

  async newItemSubmit() {
    let newItemData = {checked: false};
    let newItemInfo = this.createInput.current.value;
    if(!newItemInfo) {
      newItemInfo = '';
    }
    newItemInfo = newItemInfo.trim();
    if(newItemInfo !== '') {
      newItemData.shortDesc = newItemInfo;
      const fetchArgs = {method: 'POST',
                         body: JSON.stringify({id: this.props.id, item: newItemData}),
                         headers: {'Content-Type': 'application/json'}};
      createItem(fetchArgs)
        .then((result) => {
          newItemData.id = result.newId;
          let items = this.state.items;
          items.push(newItemData);
          this.setState({items: items});
          this.createInput.current.value = '';
        });
    }
  }

  newItemInputId() {
    return `list-${this.props.id}-newItemInput`;
  }

  render() {
    let itemBody;
    if(this.state.items.length === 0) {
      itemBody = <p>You have no items! You can add one below</p>
    } else {
      var transformedItems = this.state.items.map((item, index) => {
        return <Item key={index} id={`list-${this.props.id}-${item.id}`} shortDesc={item.shortDesc} checked={item.checked} />
      });
      itemBody = <ul>{transformedItems}</ul>
    }

    return (
      <React.Fragment>
        <h3 className="list-title">{this.props.title}</h3>
        {itemBody}
        <div className="new-item">
          <p>
            <input type="text" className="new-item-desc" placeholder="new item" ref={this.createInput} onKeyDown={this.onEnter} />
            &nbsp;
            <span className="new-item-submit" onClick={this.newItemSubmit}>+</span>
          </p>
        </div>
      </React.Fragment>
    );
  }
}

export default List;