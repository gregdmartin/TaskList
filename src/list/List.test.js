import React from 'react';
import {shallow, mount} from 'enzyme';

import List from '../list/List';
import { createItem } from '../service/ListService';
import { fail } from 'assert';

jest.mock('../service/ListService');

describe('<List />', () => {
  describe('constructor()', () => {
    it('assigns correct state values from props', () => {
      const wrapper = mount(<List id={1} title="a title" />);

      expect(wrapper.instance().createInput).toBeTruthy();
      const state = wrapper.instance().state;
      expect(state.items).toEqual([]);
    });

    it('optionally adds items from props if present', () => {
      const items = [{id: 1, shortDesc: 'an item', checked: true}];
      const wrapper = shallow(<List id={1} title="a title" items={items} />);
      const state = wrapper.instance().state;

      expect(state.items).toEqual(items);
    });
  });

  describe('onEnter()', () => {
    it('invokes newItemSubmit() if a person presses enter in the new item input', async () => {
      // make sure we have a promise to resolve in the success case
      createItem.mockImplementation(() => {
        return new Promise((resolve) => {id: 2});
      });
      const wrapper = mount(<List id="1" title="a title" />, {attachTo: document.getElementById('root')});
      const spy = jest.spyOn(wrapper.instance(), 'newItemSubmit');
      wrapper.instance().createInput.current.value = 'some text';

      const fakeEvent = {key: 'Enter'};
      await wrapper.instance().onEnter(fakeEvent);

      expect(spy).toHaveBeenCalled();
    });

    it('does nothing if the event was not from an enter keypress', async () => {
      const wrapper = mount(<List id="1" title="a title" />, {attachTo: document.getElementById('root')});
      const spy = jest.spyOn(wrapper.instance(), 'newItemSubmit');
      wrapper.instance().createInput.current.value = 'some text';

      const fakeEvent = {key: 'Lshift'};
      await wrapper.instance().onEnter(fakeEvent);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('newItemSubmit()', () => {
    it('creates a new item with the given input', async () => {
      createItem.mockImplementation(() => {
        const fakeItem = {id: 2};
        return new Promise((resolve) => resolve(fakeItem));
      });

      // we include the attachTo part to avoid attaching directly to the document body, which is considered a bad practice
      const wrapper = mount(<List id={1} title="a title"/>, {attachTo: document.getElementById('root')});
      //createInput is a ref on the List component. It's accessed as a top-level property of the wrapper
      wrapper.instance().createInput.current.value = 'some text';
      // just like componentDidMount for TaskList, this uses await to make sure api call finishes before we run expectations
      await wrapper.instance().newItemSubmit();

      const state = wrapper.instance().state;
      expect(state.items.length).toEqual(1);
    });
  });

  describe('render()', () => {
    it('renders empty when there are no items', () => {
      const wrapper = shallow(<List id={1} title="a title" />);

      expect(wrapper).toMatchSnapshot();
    });

    it('renders items when they are present', () => {
      const items = [{id: 1, shortDesc: 'an item', checked: true}];
      const wrapper = mount(<List id={1} title="a title" items={items} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});