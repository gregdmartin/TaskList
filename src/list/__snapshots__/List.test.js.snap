// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> render() renders empty when there are no items 1`] = `
<Fragment>
  <h3
    className="list-title"
  >
    a title
  </h3>
  <p>
    You have no items! You can add one below
  </p>
  <div
    className="new-item"
  >
    <p>
      <input
        className="new-item-desc"
        onKeyDown={[Function]}
        placeholder="new item"
        type="text"
      />
       
      <span
        className="new-item-submit"
        onClick={[Function]}
      >
        +
      </span>
    </p>
  </div>
</Fragment>
`;

exports[`<List /> render() renders items when they are present 1`] = `
<List
  id={1}
  items={
    Array [
      Object {
        "checked": true,
        "id": 1,
        "shortDesc": "an item",
      },
    ]
  }
  title="a title"
>
  <h3
    className="list-title"
  >
    a title
  </h3>
  <ul>
    <Item
      checked={true}
      id="list-1-1"
      key="0"
      shortDesc="an item"
    >
      <li
        className="task-item"
        id="list-1-1"
      >
        <input
          defaultChecked={true}
          onClick={[Function]}
          type="checkbox"
        />
        <span
          className="short-desc"
        >
          an item
        </span>
         
        <span
          onClick={[Function]}
        >
          [...]
        </span>
      </li>
    </Item>
  </ul>
  <div
    className="new-item"
  >
    <p>
      <input
        className="new-item-desc"
        onKeyDown={[Function]}
        placeholder="new item"
        type="text"
      />
       
      <span
        className="new-item-submit"
        onClick={[Function]}
      >
        +
      </span>
    </p>
  </div>
</List>
`;
